examples/cpp/pyperf/PyPerfBPFProgram.cc

```diff
@@ -464,9 +468,26 @@ get_thread_state(struct pt_regs *ctx) {
 
 found:
   // Get pointer to top frame from PyThreadState
-  bpf_probe_read_user(
+  if (state->offsets.PyThreadState.frame > -1) {
+    // For Python <= 3.10 get frame pointer directly from PyThreadState
+    bpf_probe_read_user(
       &state->frame_ptr, sizeof(state->frame_ptr),
       (void *)(state->thread_state + state->offsets.PyThreadState.frame));
+  } else {
+    // In Python 3.11+ PyFrameObject fields of interest were mostly moved to PyInterpreterFrame;
+    // also, we need to get pointer indirectly through PyCFrame structure
+    uintptr_t cframe;
+    bpf_probe_read_user(
+      &cframe, sizeof(cframe),
+      (void *)(state->thread_state + state->offsets.PyThreadState.cframe));
+    if (cframe == 0) {
```

Comment:
Does this mean "empty stack"? Is this state expected in any way?

(while `ThreadState->frame_ptr` can be empty, perhaps if a thread begins by running a cythonized function... what legitimacy does this state have?)

Reply:
Might not have much. Will research to make sure.

Reply:
Alright, this state doesn't appear to be possible in source code: 
- no asserts verify it, 
- no manipulations on `cframe` field are able to zero it,
- `cframe` is non-zero since initialization of `ThreadState` (checked the ThreadState init routine).

Will remove it.

Reply:
Hmm, I meant that `ERR_EMPTY_STACK` is not a good error code, and especially not very helpful because 2 different failures yield it (cframe == 0 and frame_ptr == 0).
I suggest you use another error code for it. Doing 2 consecutive yields without checking for success means that the first could fail and the second could succeed (example: page at 0x0 is mapped and so you are reading from 0x0+state->offsets.PyCFrame.current_frame).

Reply:
got it.
Will handle that case with distinct error code.